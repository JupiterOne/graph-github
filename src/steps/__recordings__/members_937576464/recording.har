{
  "log": {
    "_recordingName": "members",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "5.1.1"
    },
    "entries": [
      {
        "_id": "06d6593fc4a07c98683e9f0972b1c5e0",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/vnd.github.machine-man-preview+json"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "jupiter-integration-github octokit-rest.js/18.12.0 octokit-core.js/3.5.1 Node.js/16.14.2 (darwin; x64)"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "content-length",
              "value": "0"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.github.com"
            }
          ],
          "headersSize": 813,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "queryString": [],
          "url": "https://api.github.com/app/installations/23522163/access_tokens"
        },
        "response": {
          "bodySize": 102,
          "content": {
            "mimeType": "application/json; charset=utf-8",
            "size": 102,
            "text": "{\"message\":\"A JSON web token could not be decoded\",\"documentation_url\":\"https://docs.github.com/rest\",\"token\":\"[REDACTED]\",\"expires_at\":\"2050-12-31T18:09:20Z\"}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "server",
              "value": "GitHub.com"
            },
            {
              "name": "date",
              "value": "Fri, 02 Sep 2022 15:59:41 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=utf-8"
            },
            {
              "name": "content-length",
              "value": "102"
            },
            {
              "name": "x-github-media-type",
              "value": "github.v3; param=machine-man-preview; format=json"
            },
            {
              "name": "access-control-expose-headers",
              "value": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset"
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubdomains; preload"
            },
            {
              "name": "x-frame-options",
              "value": "deny"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "referrer-policy",
              "value": "origin-when-cross-origin, strict-origin-when-cross-origin"
            },
            {
              "name": "content-security-policy",
              "value": "default-src 'none'"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding, Accept, X-Requested-With"
            },
            {
              "name": "x-github-request-id",
              "value": "F4B9:689B:57D06:D196E:6312286D"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 939,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 401,
          "statusText": "Unauthorized"
        },
        "startedDateTime": "2022-09-02T15:59:40.909Z",
        "time": 160,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 160
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
