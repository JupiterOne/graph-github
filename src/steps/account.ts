import {
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../config';
import { createAPIClient } from '../client';
import { toAccountEntity } from '../sync/converters';
import {
  GITHUB_ACCOUNT_ENTITY_TYPE,
  GITHUB_ACCOUNT_ENTITY_CLASS,
} from '../constants';

export const DATA_ACCOUNT_ENTITY = 'DATA_ACCOUNT_ENTITY';

export async function fetchAccountDetails({
  jobState,
  instance,
  logger,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const config = instance.config;
  const apiClient = createAPIClient(config, logger);
  const accountEntity = await jobState.addEntity(
    toAccountEntity(await apiClient.getAccountDetails()),
  );

  await jobState.setData(DATA_ACCOUNT_ENTITY, accountEntity);
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-account',
    name: 'Fetch Account Details',
    entities: [
      {
        resourceName: 'Account',
        _type: GITHUB_ACCOUNT_ENTITY_TYPE,
        _class: GITHUB_ACCOUNT_ENTITY_CLASS,
      },
    ],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
