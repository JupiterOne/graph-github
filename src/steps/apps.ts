import {
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
  IntegrationMissingKeyError,
  createDirectRelationship,
} from '@jupiterone/integration-sdk-core';

import { getOrCreateApiClient } from '../client';
import { IntegrationConfig } from '../config';
import { DATA_ACCOUNT_ENTITY } from './account';
import { toAppEntity } from '../sync/converters';
import { AccountEntity, AppEntity } from '../types';
import {
  GithubEntities,
  GITHUB_ACCOUNT_APP_RELATIONSHIP_TYPE,
  GITHUB_APP_BY_APP_ID,
  Steps,
} from '../constants';

export async function fetchApps({
  instance,
  logger,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const config = instance.config;
  const apiClient = getOrCreateApiClient(config, logger);

  const accountEntity = await jobState.getData<AccountEntity>(
    DATA_ACCOUNT_ENTITY,
  );

  if (!accountEntity) {
    throw new IntegrationMissingKeyError(
      `Expected to find Account entity in jobState.`,
    );
  }

  const appIdMap = {};
  await apiClient.iterateApps(async (app) => {
    const appEntity = (await jobState.addEntity(toAppEntity(app))) as AppEntity;

    appIdMap[app.app_id] = appEntity;

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.INSTALLED,
        from: accountEntity,
        to: appEntity,
      }),
    );
  });

  await jobState.setData(GITHUB_APP_BY_APP_ID, appIdMap);
}

export const appSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: Steps.FETCH_APPS,
    name: 'Fetch Apps',
    entities: [
      {
        resourceName: 'Github App',
        _type: GithubEntities.GITHUB_APP._type,
        _class: GithubEntities.GITHUB_APP._class,
      },
    ],
    relationships: [
      {
        _type: GITHUB_ACCOUNT_APP_RELATIONSHIP_TYPE,
        _class: RelationshipClass.INSTALLED,
        sourceType: GithubEntities.GITHUB_ACCOUNT._type,
        targetType: GithubEntities.GITHUB_APP._type,
      },
    ],
    dependsOn: [Steps.FETCH_ACCOUNT],
    executionHandler: fetchApps,
  },
];
