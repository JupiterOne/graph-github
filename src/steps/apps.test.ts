import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { IntegrationConfig, sanitizeConfig } from '../config';
import { fetchApps } from './apps';
import { integrationConfig } from '../../test/config';
import { setupGithubRecording } from '../../test/recording';
import { DATA_ACCOUNT_ENTITY } from './account';
import { createMockAccountEntity } from '../../test/mockEntities';
import { GITHUB_ACCOUNT_APP_RELATIONSHIP_TYPE } from '../constants';

let recording: Recording;
afterEach(async () => {
  await recording.stop();
});

describe('fetchApps exec handler', () => {
  let apps;
  let context;

  test('execution and snapshot', async () => {
    recording = setupGithubRecording({
      directory: __dirname,
      name: 'apps', //redaction of headers is in setupGithubRecording
    });
    sanitizeConfig(integrationConfig);
    integrationConfig.installationId = 17214088; //this is the id the recordings are under
    const accountEntity = createMockAccountEntity();
    context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
      setData: {
        [DATA_ACCOUNT_ENTITY]: accountEntity,
      },
    });

    // Simulates dependency graph execution.
    // See https://github.com/JupiterOne/sdk/issues/262.
    await fetchApps(context);

    // Review snapshot, failure is a regression
    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    apps = context.jobState.collectedEntities.filter((e) =>
      e._class.includes('Application'),
    );
    expect(apps.length).toBeGreaterThan(0);

    const relationships = context.jobState.collectedRelationships.filter((e) =>
      e._type.includes(GITHUB_ACCOUNT_APP_RELATIONSHIP_TYPE),
    );
    expect(relationships.length).toBeGreaterThan(0);
  });

  test('schema match', () => {
    expect(apps).toMatchGraphObjectSchema({
      _class: ['Application'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'github_app' },
          name: { type: 'string' },
          displayName: { type: 'string' },
          webLink: { type: 'string' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
        },
        required: ['name', 'displayName', 'webLink', 'createdOn'],
      },
    });
  });
});
