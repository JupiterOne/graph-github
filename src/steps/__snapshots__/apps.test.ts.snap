// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchApps exec handler 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "GitHubAppInstallation_23522163",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "access_tokens_url": "https://api.github.com/app/installations/23522163/access_tokens",
            "account": Object {
              "avatar_url": "https://avatars.githubusercontent.com/u/100235621?v=4",
              "events_url": "https://api.github.com/users/j1-ingest/events{/privacy}",
              "followers_url": "https://api.github.com/users/j1-ingest/followers",
              "following_url": "https://api.github.com/users/j1-ingest/following{/other_user}",
              "gists_url": "https://api.github.com/users/j1-ingest/gists{/gist_id}",
              "gravatar_id": "",
              "html_url": "https://github.com/j1-ingest",
              "id": 100235621,
              "login": "j1-ingest",
              "node_id": "O_kgDOBfl5ZQ",
              "organizations_url": "https://api.github.com/users/j1-ingest/orgs",
              "received_events_url": "https://api.github.com/users/j1-ingest/received_events",
              "repos_url": "https://api.github.com/users/j1-ingest/repos",
              "site_admin": false,
              "starred_url": "https://api.github.com/users/j1-ingest/starred{/owner}{/repo}",
              "subscriptions_url": "https://api.github.com/users/j1-ingest/subscriptions",
              "type": "Organization",
              "url": "https://api.github.com/users/j1-ingest",
            },
            "app_id": 174703,
            "app_slug": "jupiterone-spoulton",
            "created_at": "2022-02-22T19:50:45.000Z",
            "events": Array [],
            "has_multiple_single_files": false,
            "html_url": "https://github.com/organizations/j1-ingest/settings/installations/23522163",
            "id": 23522163,
            "permissions": Object {
              "administration": "read",
              "environments": "read",
              "issues": "read",
              "members": "read",
              "metadata": "read",
              "organization_administration": "read",
              "organization_secrets": "read",
              "pull_requests": "read",
              "secrets": "read",
              "vulnerability_alerts": "read",
            },
            "repositories_url": "https://api.github.com/installation/repositories",
            "repository_selection": "all",
            "single_file_name": null,
            "single_file_paths": Array [],
            "suspended_at": null,
            "suspended_by": null,
            "target_id": 100235621,
            "target_type": "Organization",
            "updated_at": "2022-05-19T16:05:01.000Z",
          },
        },
      ],
      "_type": "github_app",
      "appId": 174703,
      "appSlug": "jupiterone-spoulton",
      "createdOn": 1645559445000,
      "displayName": "jupiterone-spoulton",
      "events": Array [],
      "hasMultipleSingleFiles": false,
      "installationId": 23522163,
      "name": "jupiterone-spoulton",
      "permissions.administration": "read",
      "permissions.environments": "read",
      "permissions.issues": "read",
      "permissions.members": "read",
      "permissions.metadata": "read",
      "permissions.organization-administration": "read",
      "permissions.organization-secrets": "read",
      "permissions.pull-requests": "read",
      "permissions.secrets": "read",
      "permissions.vulnerability-alerts": "read",
      "repositorySelection": "all",
      "singleFileName": "",
      "singleFilePaths": Array [],
      "suspendedOn": undefined,
      "targetId": 100235621,
      "targetType": "Organization",
      "updatedOn": 1652976301000,
      "webLink": "https://github.com/organizations/j1-ingest/settings/installations/23522163",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "GitHubAppInstallation_23790442",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "access_tokens_url": "https://api.github.com/app/installations/23790442/access_tokens",
            "account": Object {
              "avatar_url": "https://avatars.githubusercontent.com/u/100235621?v=4",
              "events_url": "https://api.github.com/users/j1-ingest/events{/privacy}",
              "followers_url": "https://api.github.com/users/j1-ingest/followers",
              "following_url": "https://api.github.com/users/j1-ingest/following{/other_user}",
              "gists_url": "https://api.github.com/users/j1-ingest/gists{/gist_id}",
              "gravatar_id": "",
              "html_url": "https://github.com/j1-ingest",
              "id": 100235621,
              "login": "j1-ingest",
              "node_id": "O_kgDOBfl5ZQ",
              "organizations_url": "https://api.github.com/users/j1-ingest/orgs",
              "received_events_url": "https://api.github.com/users/j1-ingest/received_events",
              "repos_url": "https://api.github.com/users/j1-ingest/repos",
              "site_admin": false,
              "starred_url": "https://api.github.com/users/j1-ingest/starred{/owner}{/repo}",
              "subscriptions_url": "https://api.github.com/users/j1-ingest/subscriptions",
              "type": "Organization",
              "url": "https://api.github.com/users/j1-ingest",
            },
            "app_id": 147668,
            "app_slug": "cisco-secure-cloud-insights-dev",
            "created_at": "2022-03-03T18:42:50.000Z",
            "events": Array [],
            "has_multiple_single_files": false,
            "html_url": "https://github.com/organizations/j1-ingest/settings/installations/23790442",
            "id": 23790442,
            "permissions": Object {
              "actions": "read",
              "environments": "read",
              "members": "read",
              "metadata": "read",
              "organization_administration": "read",
              "organization_secrets": "read",
              "pull_requests": "read",
              "secrets": "read",
            },
            "repositories_url": "https://api.github.com/installation/repositories",
            "repository_selection": "all",
            "single_file_name": null,
            "single_file_paths": Array [],
            "suspended_at": null,
            "suspended_by": null,
            "target_id": 100235621,
            "target_type": "Organization",
            "updated_at": "2022-03-03T18:42:50.000Z",
          },
        },
      ],
      "_type": "github_app",
      "appId": 147668,
      "appSlug": "cisco-secure-cloud-insights-dev",
      "createdOn": 1646332970000,
      "displayName": "cisco-secure-cloud-insights-dev",
      "events": Array [],
      "hasMultipleSingleFiles": false,
      "installationId": 23790442,
      "name": "cisco-secure-cloud-insights-dev",
      "permissions.actions": "read",
      "permissions.environments": "read",
      "permissions.members": "read",
      "permissions.metadata": "read",
      "permissions.organization-administration": "read",
      "permissions.organization-secrets": "read",
      "permissions.pull-requests": "read",
      "permissions.secrets": "read",
      "repositorySelection": "all",
      "singleFileName": "",
      "singleFilePaths": Array [],
      "suspendedOn": undefined,
      "targetId": 100235621,
      "targetType": "Organization",
      "updatedOn": 1646332970000,
      "webLink": "https://github.com/organizations/j1-ingest/settings/installations/23790442",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "GitHubAppInstallation_27135279",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "access_tokens_url": "https://api.github.com/app/installations/27135279/access_tokens",
            "account": Object {
              "avatar_url": "https://avatars.githubusercontent.com/u/100235621?v=4",
              "events_url": "https://api.github.com/users/j1-ingest/events{/privacy}",
              "followers_url": "https://api.github.com/users/j1-ingest/followers",
              "following_url": "https://api.github.com/users/j1-ingest/following{/other_user}",
              "gists_url": "https://api.github.com/users/j1-ingest/gists{/gist_id}",
              "gravatar_id": "",
              "html_url": "https://github.com/j1-ingest",
              "id": 100235621,
              "login": "j1-ingest",
              "node_id": "O_kgDOBfl5ZQ",
              "organizations_url": "https://api.github.com/users/j1-ingest/orgs",
              "received_events_url": "https://api.github.com/users/j1-ingest/received_events",
              "repos_url": "https://api.github.com/users/j1-ingest/repos",
              "site_admin": false,
              "starred_url": "https://api.github.com/users/j1-ingest/starred{/owner}{/repo}",
              "subscriptions_url": "https://api.github.com/users/j1-ingest/subscriptions",
              "type": "Organization",
              "url": "https://api.github.com/users/j1-ingest",
            },
            "app_id": 19640,
            "app_slug": "jupiterone-dev",
            "created_at": "2022-07-05T19:03:29.000Z",
            "events": Array [],
            "has_multiple_single_files": false,
            "html_url": "https://github.com/organizations/j1-ingest/settings/installations/27135279",
            "id": 27135279,
            "permissions": Object {
              "actions": "read",
              "environments": "read",
              "issues": "read",
              "members": "read",
              "metadata": "read",
              "organization_administration": "read",
              "organization_secrets": "read",
              "pull_requests": "read",
              "secrets": "read",
              "statuses": "read",
              "vulnerability_alerts": "read",
            },
            "repositories_url": "https://api.github.com/installation/repositories",
            "repository_selection": "all",
            "single_file_name": null,
            "single_file_paths": Array [],
            "suspended_at": null,
            "suspended_by": null,
            "target_id": 100235621,
            "target_type": "Organization",
            "updated_at": "2022-07-05T19:03:29.000Z",
          },
        },
      ],
      "_type": "github_app",
      "appId": 19640,
      "appSlug": "jupiterone-dev",
      "createdOn": 1657047809000,
      "displayName": "jupiterone-dev",
      "events": Array [],
      "hasMultipleSingleFiles": false,
      "installationId": 27135279,
      "name": "jupiterone-dev",
      "permissions.actions": "read",
      "permissions.environments": "read",
      "permissions.issues": "read",
      "permissions.members": "read",
      "permissions.metadata": "read",
      "permissions.organization-administration": "read",
      "permissions.organization-secrets": "read",
      "permissions.pull-requests": "read",
      "permissions.secrets": "read",
      "permissions.statuses": "read",
      "permissions.vulnerability-alerts": "read",
      "repositorySelection": "all",
      "singleFileName": "",
      "singleFilePaths": Array [],
      "suspendedOn": undefined,
      "targetId": 100235621,
      "targetType": "Organization",
      "updatedOn": 1657047809000,
      "webLink": "https://github.com/organizations/j1-ingest/settings/installations/27135279",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "GitHubAppInstallation_27613673",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "access_tokens_url": "https://api.github.com/app/installations/27613673/access_tokens",
            "account": Object {
              "avatar_url": "https://avatars.githubusercontent.com/u/100235621?v=4",
              "events_url": "https://api.github.com/users/j1-ingest/events{/privacy}",
              "followers_url": "https://api.github.com/users/j1-ingest/followers",
              "following_url": "https://api.github.com/users/j1-ingest/following{/other_user}",
              "gists_url": "https://api.github.com/users/j1-ingest/gists{/gist_id}",
              "gravatar_id": "",
              "html_url": "https://github.com/j1-ingest",
              "id": 100235621,
              "login": "j1-ingest",
              "node_id": "O_kgDOBfl5ZQ",
              "organizations_url": "https://api.github.com/users/j1-ingest/orgs",
              "received_events_url": "https://api.github.com/users/j1-ingest/received_events",
              "repos_url": "https://api.github.com/users/j1-ingest/repos",
              "site_admin": false,
              "starred_url": "https://api.github.com/users/j1-ingest/starred{/owner}{/repo}",
              "subscriptions_url": "https://api.github.com/users/j1-ingest/subscriptions",
              "type": "Organization",
              "url": "https://api.github.com/users/j1-ingest",
            },
            "app_id": 222078,
            "app_slug": "full-ingest-from-vdubber",
            "created_at": "2022-07-22T16:32:16.000Z",
            "events": Array [],
            "has_multiple_single_files": true,
            "html_url": "https://github.com/organizations/j1-ingest/settings/installations/27613673",
            "id": 27613673,
            "permissions": Object {
              "administration": "write",
              "members": "read",
              "metadata": "read",
              "organization_administration": "write",
              "organization_dependabot_secrets": "read",
              "organization_events": "read",
              "organization_hooks": "read",
              "organization_plan": "read",
              "organization_projects": "read",
              "organization_secrets": "read",
              "organization_self_hosted_runners": "read",
              "organization_user_blocking": "read",
              "single_file": "read",
              "team_discussions": "read",
            },
            "repositories_url": "https://api.github.com/installation/repositories",
            "repository_selection": "all",
            "single_file_name": "CODEOWNER",
            "single_file_paths": Array [
              "CODEOWNER",
              "package.json",
              "package.lock",
              "yarn.lock",
            ],
            "suspended_at": null,
            "suspended_by": null,
            "target_id": 100235621,
            "target_type": "Organization",
            "updated_at": "2022-07-22T16:40:54.000Z",
          },
        },
      ],
      "_type": "github_app",
      "appId": 222078,
      "appSlug": "full-ingest-from-vdubber",
      "createdOn": 1658507536000,
      "displayName": "full-ingest-from-vdubber",
      "events": Array [],
      "hasMultipleSingleFiles": true,
      "installationId": 27613673,
      "name": "full-ingest-from-vdubber",
      "permissions.administration": "write",
      "permissions.members": "read",
      "permissions.metadata": "read",
      "permissions.organization-administration": "write",
      "permissions.organization-dependabot-secrets": "read",
      "permissions.organization-events": "read",
      "permissions.organization-hooks": "read",
      "permissions.organization-plan": "read",
      "permissions.organization-projects": "read",
      "permissions.organization-secrets": "read",
      "permissions.organization-self-hosted-runners": "read",
      "permissions.organization-user-blocking": "read",
      "permissions.single-file": "read",
      "permissions.team-discussions": "read",
      "repositorySelection": "all",
      "singleFileName": "CODEOWNER",
      "singleFilePaths": Array [
        "CODEOWNER",
        "package.json",
        "package.lock",
        "yarn.lock",
      ],
      "suspendedOn": undefined,
      "targetId": 100235621,
      "targetType": "Organization",
      "updatedOn": 1658508054000,
      "webLink": "https://github.com/organizations/j1-ingest/settings/installations/27613673",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "GitHubAppInstallation_27915740",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "access_tokens_url": "https://api.github.com/app/installations/27915740/access_tokens",
            "account": Object {
              "avatar_url": "https://avatars.githubusercontent.com/u/100235621?v=4",
              "events_url": "https://api.github.com/users/j1-ingest/events{/privacy}",
              "followers_url": "https://api.github.com/users/j1-ingest/followers",
              "following_url": "https://api.github.com/users/j1-ingest/following{/other_user}",
              "gists_url": "https://api.github.com/users/j1-ingest/gists{/gist_id}",
              "gravatar_id": "",
              "html_url": "https://github.com/j1-ingest",
              "id": 100235621,
              "login": "j1-ingest",
              "node_id": "O_kgDOBfl5ZQ",
              "organizations_url": "https://api.github.com/users/j1-ingest/orgs",
              "received_events_url": "https://api.github.com/users/j1-ingest/received_events",
              "repos_url": "https://api.github.com/users/j1-ingest/repos",
              "site_admin": false,
              "starred_url": "https://api.github.com/users/j1-ingest/starred{/owner}{/repo}",
              "subscriptions_url": "https://api.github.com/users/j1-ingest/subscriptions",
              "type": "Organization",
              "url": "https://api.github.com/users/j1-ingest",
            },
            "app_id": 20361,
            "app_slug": "jupiterone-io",
            "created_at": "2022-08-02T21:01:24.000Z",
            "events": Array [],
            "has_multiple_single_files": false,
            "html_url": "https://github.com/organizations/j1-ingest/settings/installations/27915740",
            "id": 27915740,
            "permissions": Object {
              "actions": "read",
              "environments": "read",
              "issues": "read",
              "members": "read",
              "metadata": "read",
              "organization_administration": "read",
              "organization_secrets": "read",
              "pull_requests": "read",
              "secrets": "read",
              "vulnerability_alerts": "read",
            },
            "repositories_url": "https://api.github.com/installation/repositories",
            "repository_selection": "all",
            "single_file_name": null,
            "single_file_paths": Array [],
            "suspended_at": null,
            "suspended_by": null,
            "target_id": 100235621,
            "target_type": "Organization",
            "updated_at": "2022-08-02T21:01:24.000Z",
          },
        },
      ],
      "_type": "github_app",
      "appId": 20361,
      "appSlug": "jupiterone-io",
      "createdOn": 1659474084000,
      "displayName": "jupiterone-io",
      "events": Array [],
      "hasMultipleSingleFiles": false,
      "installationId": 27915740,
      "name": "jupiterone-io",
      "permissions.actions": "read",
      "permissions.environments": "read",
      "permissions.issues": "read",
      "permissions.members": "read",
      "permissions.metadata": "read",
      "permissions.organization-administration": "read",
      "permissions.organization-secrets": "read",
      "permissions.pull-requests": "read",
      "permissions.secrets": "read",
      "permissions.vulnerability-alerts": "read",
      "repositorySelection": "all",
      "singleFileName": "",
      "singleFilePaths": Array [],
      "suspendedOn": undefined,
      "targetId": 100235621,
      "targetType": "Organization",
      "updatedOn": 1659474084000,
      "webLink": "https://github.com/organizations/j1-ingest/settings/installations/27915740",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "GitHubAppInstallation_28077570",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "access_tokens_url": "https://api.github.com/app/installations/28077570/access_tokens",
            "account": Object {
              "avatar_url": "https://avatars.githubusercontent.com/u/100235621?v=4",
              "events_url": "https://api.github.com/users/j1-ingest/events{/privacy}",
              "followers_url": "https://api.github.com/users/j1-ingest/followers",
              "following_url": "https://api.github.com/users/j1-ingest/following{/other_user}",
              "gists_url": "https://api.github.com/users/j1-ingest/gists{/gist_id}",
              "gravatar_id": "",
              "html_url": "https://github.com/j1-ingest",
              "id": 100235621,
              "login": "j1-ingest",
              "node_id": "O_kgDOBfl5ZQ",
              "organizations_url": "https://api.github.com/users/j1-ingest/orgs",
              "received_events_url": "https://api.github.com/users/j1-ingest/received_events",
              "repos_url": "https://api.github.com/users/j1-ingest/repos",
              "site_admin": false,
              "starred_url": "https://api.github.com/users/j1-ingest/starred{/owner}{/repo}",
              "subscriptions_url": "https://api.github.com/users/j1-ingest/subscriptions",
              "type": "Organization",
              "url": "https://api.github.com/users/j1-ingest",
            },
            "app_id": 226546,
            "app_slug": "adaminict-test",
            "created_at": "2022-08-08T18:43:57.000Z",
            "events": Array [],
            "has_multiple_single_files": true,
            "html_url": "https://github.com/organizations/j1-ingest/settings/installations/28077570",
            "id": 28077570,
            "permissions": Object {
              "actions": "read",
              "administration": "write",
              "contents": "read",
              "environments": "read",
              "issues": "read",
              "members": "read",
              "metadata": "read",
              "organization_administration": "write",
              "organization_dependabot_secrets": "read",
              "organization_hooks": "read",
              "organization_plan": "read",
              "organization_projects": "read",
              "organization_secrets": "read",
              "organization_self_hosted_runners": "read",
              "organization_user_blocking": "read",
              "pull_requests": "read",
              "secrets": "read",
              "single_file": "read",
              "team_discussions": "read",
              "vulnerability_alerts": "read",
            },
            "repositories_url": "https://api.github.com/installation/repositories",
            "repository_selection": "all",
            "single_file_name": "CODEOWNER",
            "single_file_paths": Array [
              "CODEOWNER",
              "package.json",
              "package.lock",
              "yarn.lock",
            ],
            "suspended_at": null,
            "suspended_by": null,
            "target_id": 100235621,
            "target_type": "Organization",
            "updated_at": "2022-08-08T18:52:45.000Z",
          },
        },
      ],
      "_type": "github_app",
      "appId": 226546,
      "appSlug": "adaminict-test",
      "createdOn": 1659984237000,
      "displayName": "adaminict-test",
      "events": Array [],
      "hasMultipleSingleFiles": true,
      "installationId": 28077570,
      "name": "adaminict-test",
      "permissions.actions": "read",
      "permissions.administration": "write",
      "permissions.contents": "read",
      "permissions.environments": "read",
      "permissions.issues": "read",
      "permissions.members": "read",
      "permissions.metadata": "read",
      "permissions.organization-administration": "write",
      "permissions.organization-dependabot-secrets": "read",
      "permissions.organization-hooks": "read",
      "permissions.organization-plan": "read",
      "permissions.organization-projects": "read",
      "permissions.organization-secrets": "read",
      "permissions.organization-self-hosted-runners": "read",
      "permissions.organization-user-blocking": "read",
      "permissions.pull-requests": "read",
      "permissions.secrets": "read",
      "permissions.single-file": "read",
      "permissions.team-discussions": "read",
      "permissions.vulnerability-alerts": "read",
      "repositorySelection": "all",
      "singleFileName": "CODEOWNER",
      "singleFilePaths": Array [
        "CODEOWNER",
        "package.json",
        "package.lock",
        "yarn.lock",
      ],
      "suspendedOn": undefined,
      "targetId": 100235621,
      "targetType": "Organization",
      "updatedOn": 1659984765000,
      "webLink": "https://github.com/organizations/j1-ingest/settings/installations/28077570",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "INSTALLED",
      "_fromEntityKey": "O_kgDOBfl5ZQ",
      "_key": "O_kgDOBfl5ZQ|installed|GitHubAppInstallation_23522163",
      "_toEntityKey": "GitHubAppInstallation_23522163",
      "_type": "github_account_installed_app",
      "displayName": "INSTALLED",
    },
    Object {
      "_class": "INSTALLED",
      "_fromEntityKey": "O_kgDOBfl5ZQ",
      "_key": "O_kgDOBfl5ZQ|installed|GitHubAppInstallation_23790442",
      "_toEntityKey": "GitHubAppInstallation_23790442",
      "_type": "github_account_installed_app",
      "displayName": "INSTALLED",
    },
    Object {
      "_class": "INSTALLED",
      "_fromEntityKey": "O_kgDOBfl5ZQ",
      "_key": "O_kgDOBfl5ZQ|installed|GitHubAppInstallation_27135279",
      "_toEntityKey": "GitHubAppInstallation_27135279",
      "_type": "github_account_installed_app",
      "displayName": "INSTALLED",
    },
    Object {
      "_class": "INSTALLED",
      "_fromEntityKey": "O_kgDOBfl5ZQ",
      "_key": "O_kgDOBfl5ZQ|installed|GitHubAppInstallation_27613673",
      "_toEntityKey": "GitHubAppInstallation_27613673",
      "_type": "github_account_installed_app",
      "displayName": "INSTALLED",
    },
    Object {
      "_class": "INSTALLED",
      "_fromEntityKey": "O_kgDOBfl5ZQ",
      "_key": "O_kgDOBfl5ZQ|installed|GitHubAppInstallation_27915740",
      "_toEntityKey": "GitHubAppInstallation_27915740",
      "_type": "github_account_installed_app",
      "displayName": "INSTALLED",
    },
    Object {
      "_class": "INSTALLED",
      "_fromEntityKey": "O_kgDOBfl5ZQ",
      "_key": "O_kgDOBfl5ZQ|installed|GitHubAppInstallation_28077570",
      "_toEntityKey": "GitHubAppInstallation_28077570",
      "_type": "github_account_installed_app",
      "displayName": "INSTALLED",
    },
  ],
  "encounteredTypes": Array [
    "github_account",
    "github_app",
    "github_account_installed_app",
  ],
  "numCollectedEntities": 6,
  "numCollectedRelationships": 6,
}
`;
